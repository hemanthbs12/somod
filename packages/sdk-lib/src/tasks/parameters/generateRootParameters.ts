export { generate as generateRootParameters } from "../../utils/parameters/generate";
