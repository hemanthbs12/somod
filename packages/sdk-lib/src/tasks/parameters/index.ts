export { validateSchema as validateParametersWithSchema } from "./validateSchema";
export { buildParameters } from "./buildParameters";
export { generateRootParameters } from "./generateRootParameters";
