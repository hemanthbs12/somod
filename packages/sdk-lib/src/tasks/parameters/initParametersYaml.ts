import { existsSync } from "fs";
import { mkdir, writeFile } from "fs/promises";
import { dirname, join } from "path";
import { file_parametersYaml } from "../../utils/constants";

export const initParametersYaml = async (dir: string): Promise<void> => {
  const parametersPath = join(dir, file_parametersYaml);

  const parametersContent = `# yaml-language-server: $schema=./node_modules/@somod/parameters-schema/schemas/index.json

Parameters: {}

`;

  if (!existsSync(parametersPath)) {
    const templateDir = dirname(parametersPath);
    await mkdir(templateDir, { recursive: true });
    await writeFile(parametersPath, parametersContent);
  }
};
