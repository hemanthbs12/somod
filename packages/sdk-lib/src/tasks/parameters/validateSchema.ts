import { join } from "path";
import {
  defaultParametersSchema,
  file_parametersYaml
} from "../../utils/constants";
import { yamlSchemaValidator } from "../../utils/yamlSchemaValidator";

export const validateSchema = async (dir: string): Promise<void> => {
  await yamlSchemaValidator(
    defaultParametersSchema,
    join(dir, file_parametersYaml),
    dir
  );
};
