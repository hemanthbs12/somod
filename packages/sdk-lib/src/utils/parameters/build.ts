import { mkdir, writeFile } from "fs/promises";
import { dirname, join } from "path";
import {
  file_parametersJson,
  file_parametersYaml,
  path_build
} from "../constants";
import { ModuleHandler, ModuleNode } from "../moduleHandler";
import { readYamlFileStore } from "../yamlFileStore";
import { loadParameters } from "./load";

const buildParametersYaml = async (dir: string): Promise<void> => {
  const parameters = await readYamlFileStore(join(dir, file_parametersYaml));
  const parametersBuildPath = join(dir, path_build, file_parametersJson);
  await mkdir(dirname(parametersBuildPath), { recursive: true });
  await writeFile(parametersBuildPath, JSON.stringify(parameters));
};

const validate = async (moduleNode: ModuleNode): Promise<void> => {
  const parameters = await loadParameters(moduleNode.module);

  const parametersWithLongName = Object.keys(
    parameters.Parameters || {}
  ).filter(parameterName => parameterName.length > 128);

  if (parametersWithLongName.length > 0) {
    throw new Error(
      `Following parameters have too long name (maximum 128 characters are allowed)\n${parametersWithLongName
        .map(p => " - " + p)
        .join("\n")}`
    );
  }
};

export const build = async (
  dir: string,
  moduleIndicators: string[]
): Promise<void> => {
  const moduleHandler = ModuleHandler.getModuleHandler(dir, moduleIndicators);

  const rootModuleNode = await moduleHandler.getRoodModuleNode();

  await validate(rootModuleNode);

  await buildParametersYaml(dir);
};
