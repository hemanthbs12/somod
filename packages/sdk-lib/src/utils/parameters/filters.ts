import { JSONSchema7, getCompiledValidator } from "@solib/json-validator";
import { difference } from "lodash";

export type FilterFunction = (ip: unknown) => unknown | Promise<unknown>;

export class Filter {
  private static filter: Filter;

  private filters: Record<string, FilterFunction> = {};

  private constructor() {
    // dont do anything
  }

  public static getFilter() {
    if (!this.filter) {
      this.filter = new Filter();
      this.filter.register("ajvStandalone", ajvStandalone);
      this.filter.register("jsonParse", jsonParse);
      this.filter.register("jsonStringify", jsonStringify);
    }
    return this.filter;
  }

  public register(name: string, fn: FilterFunction) {
    this.filters[name] = fn;
  }

  public async apply(ip: unknown, filterNames: string[]): Promise<unknown> {
    const registeredFilterNames = Object.keys(this.filters);
    const unknownFilterNames = difference(filterNames, registeredFilterNames);
    if (unknownFilterNames.length > 0) {
      throw new Error(
        `Unknown Parameters filters (${unknownFilterNames.join(", ")})`
      );
    }

    let intermediate = ip;
    for (const filterName of filterNames) {
      const result = this.filters[filterName](intermediate);
      intermediate =
        result && typeof (result as Promise<unknown>).then == "function"
          ? await result
          : result;
    }

    return intermediate;
  }
}

export const ajvStandalone = (schema: JSONSchema7) => {
  return getCompiledValidator(schema);
};

export const jsonParse = (json: string) => {
  return JSON.parse(json);
};

export const jsonStringify = (json: unknown) => {
  return JSON.stringify(json);
};
