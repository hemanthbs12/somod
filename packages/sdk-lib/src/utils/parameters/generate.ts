import {
  readJsonFileStore,
  saveJsonFileStore,
  updateJsonFileStore
} from "@solib/cli-base";
import { existsSync } from "fs";
import { uniq } from "lodash";
import { join } from "path";
import { file_parametersJson, namespace_parameter } from "../constants";
import { ModuleHandler } from "../moduleHandler";
import { loadParameters } from "./load";
import { loadParameterNamespaces } from "./namespace";
import { Parameters } from "./types";

/**
 * Creates/Updates `parameters.json` at the root of the project directory.
 * `parameters.json` contains map of parameter name to default value of the parameter
 *
 * if **`oveerride`** is **true** all the existing values in parameters.json are overrided by default value, otherwise only new parameters are added to `parameters.json` if any
 */
export const generate = async (
  dir: string,
  moduleIndicators: string[],
  override = false
): Promise<void> => {
  const moduleHandler = ModuleHandler.getModuleHandler(dir, moduleIndicators);

  const namespaces = await moduleHandler.getNamespaces(
    Object.fromEntries(
      moduleIndicators.map(mt => [mt, loadParameterNamespaces])
    )
  );

  const parameterToModuleNameMap = namespaces[namespace_parameter];

  const allModuleNames = uniq(Object.values(parameterToModuleNameMap));

  const moduleToParametersMap: Record<string, Parameters> = {};

  await Promise.all(
    allModuleNames.map(async moduleName => {
      const moduleNode = await moduleHandler.getModule(moduleName);
      moduleToParametersMap[moduleName] = await loadParameters(
        moduleNode.module
      );
    })
  );

  const parameterDefaultValues = Object.fromEntries(
    Object.keys(parameterToModuleNameMap).map(parameterName => [
      parameterName,
      moduleToParametersMap[parameterToModuleNameMap[parameterName]].Parameters[
        parameterName
      ].default
    ])
  );

  const rootParametersJsonPath = join(dir, file_parametersJson);
  const existingParameters = existsSync(rootParametersJsonPath)
    ? await readJsonFileStore(rootParametersJsonPath)
    : {};

  const newParameters = override
    ? { ...existingParameters, ...parameterDefaultValues }
    : { ...parameterDefaultValues, ...existingParameters };

  updateJsonFileStore(rootParametersJsonPath, newParameters);
  await saveJsonFileStore(rootParametersJsonPath);
};
