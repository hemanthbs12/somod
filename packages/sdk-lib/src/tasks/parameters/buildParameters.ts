import { existsSync } from "fs";
import { join } from "path";
import { file_parametersYaml } from "../../utils/constants";
import { build } from "../../utils/parameters/build";

export const buildParameters = async (
  dir: string,
  moduleIndicators: string[]
): Promise<void> => {
  if (existsSync(join(dir, file_parametersYaml))) {
    await build(dir, moduleIndicators);
  }
};
