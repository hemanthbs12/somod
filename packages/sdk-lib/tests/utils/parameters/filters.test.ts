import { Filter } from "../../../src/utils/parameters/filters";

describe("Test util parameters.filters", () => {
  test("default filters", () => {
    const filter = Filter.getFilter();

    expect(Object.keys(filter["filters"])).toEqual([
      "ajvStandalone",
      "jsonParse",
      "jsonStringify"
    ]);
  });

  test("with additional filters", () => {
    const filter = Filter.getFilter();
    filter.register("customFilter", (ip: string) => {
      return ip + ip;
    });

    expect(Object.keys(filter["filters"])).toEqual([
      "ajvStandalone",
      "jsonParse",
      "jsonStringify",
      "customFilter"
    ]);
  });

  test("with apply ajvStandalone", async () => {
    const filter = Filter.getFilter();

    await expect(
      filter.apply(
        {
          type: "object",
          properties: {
            name: { type: "string" }
          }
        },
        ["ajvStandalone"]
      )
    ).resolves.toMatchSnapshot();
  });

  test("with apply multiple filters", async () => {
    const filter = Filter.getFilter();
    filter.register("customFilter", (ip: string) => {
      return ip + ip;
    });

    await expect(
      filter.apply(
        {
          type: "object",
          properties: {
            name: { type: "string" }
          }
        },
        ["jsonStringify", "jsonParse", "ajvStandalone", "customFilter"]
      )
    ).resolves.toMatchSnapshot();
  });
});
