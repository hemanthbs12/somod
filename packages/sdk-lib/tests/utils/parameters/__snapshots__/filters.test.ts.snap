// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test util parameters.filters with apply ajvStandalone 1`] = `"\\"use strict\\";module.exports = validate14;module.exports.default = validate14;const schema16 = {\\"type\\":\\"object\\",\\"properties\\":{\\"name\\":{\\"type\\":\\"string\\"}}};function validate14(data, {instancePath=\\"\\", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(data && typeof data == \\"object\\" && !Array.isArray(data)){if(data.name !== undefined){if(typeof data.name !== \\"string\\"){const err0 = {instancePath:instancePath+\\"/name\\",schemaPath:\\"#/properties/name/type\\",keyword:\\"type\\",params:{type: \\"string\\"},message:\\"must be string\\"};if(vErrors === null){vErrors = [err0];}else {vErrors.push(err0);}errors++;}}}else {const err1 = {instancePath,schemaPath:\\"#/type\\",keyword:\\"type\\",params:{type: \\"object\\"},message:\\"must be object\\"};if(vErrors === null){vErrors = [err1];}else {vErrors.push(err1);}errors++;}validate14.errors = vErrors;return errors === 0;}"`;

exports[`Test util parameters.filters with apply multiple filters 1`] = `"\\"use strict\\";module.exports = validate14;module.exports.default = validate14;const schema16 = {\\"type\\":\\"object\\",\\"properties\\":{\\"name\\":{\\"type\\":\\"string\\"}}};function validate14(data, {instancePath=\\"\\", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(data && typeof data == \\"object\\" && !Array.isArray(data)){if(data.name !== undefined){if(typeof data.name !== \\"string\\"){const err0 = {instancePath:instancePath+\\"/name\\",schemaPath:\\"#/properties/name/type\\",keyword:\\"type\\",params:{type: \\"string\\"},message:\\"must be string\\"};if(vErrors === null){vErrors = [err0];}else {vErrors.push(err0);}errors++;}}}else {const err1 = {instancePath,schemaPath:\\"#/type\\",keyword:\\"type\\",params:{type: \\"object\\"},message:\\"must be object\\"};if(vErrors === null){vErrors = [err1];}else {vErrors.push(err1);}errors++;}validate14.errors = vErrors;return errors === 0;}\\"use strict\\";module.exports = validate14;module.exports.default = validate14;const schema16 = {\\"type\\":\\"object\\",\\"properties\\":{\\"name\\":{\\"type\\":\\"string\\"}}};function validate14(data, {instancePath=\\"\\", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(data && typeof data == \\"object\\" && !Array.isArray(data)){if(data.name !== undefined){if(typeof data.name !== \\"string\\"){const err0 = {instancePath:instancePath+\\"/name\\",schemaPath:\\"#/properties/name/type\\",keyword:\\"type\\",params:{type: \\"string\\"},message:\\"must be string\\"};if(vErrors === null){vErrors = [err0];}else {vErrors.push(err0);}errors++;}}}else {const err1 = {instancePath,schemaPath:\\"#/type\\",keyword:\\"type\\",params:{type: \\"object\\"},message:\\"must be object\\"};if(vErrors === null){vErrors = [err1];}else {vErrors.push(err1);}errors++;}validate14.errors = vErrors;return errors === 0;}"`;
